File: packages/api/src/examples/resilienceExample.ts
/**
 * Example usage of resilience patterns (retry and circuit breaker).
 * 
 * This module demonstrates how to use retry logic and circuit breakers
 * to handle failures in external service calls.
 */
import {
  retry,
  RetryStrategy,
  CircuitBreaker,
  withCircuitBreaker,
  withFallback,
  withTimeout,
  registerCircuitBreaker,
} from '../resilience';
import { DataSourceError, ErrorCode } from '../errors';
import { logger } from '../logging';

// Example 1: Using retry function with exponential backoff
async function fetchDataFromAPI(url: string): Promise<Record<string, any>> {
  logger.info(`Fetching data from ${url}`);

  // Simulate API call that might fail
  if (Math.random() < 0.3) {
    // 30% chance of failure
    throw new DataSourceError(
      'API temporarily unavailable',
      url,
      ErrorCode.DATA_SOURCE_UNAVAILABLE,
      true,
    );
  }

  return { status: 'success', data: [1, 2, 3] };
}

async function fetchWithRetry(url: string): Promise<Record<string, any>> {
  return retry(() => fetchDataFromAPI(url), {
    maxAttempts: 3,
    strategy: RetryStrategy.EXPONENTIAL,
    baseDelay: 1000,
  });
}

// Example 2: Using circuit breaker class
const exchangeBreaker = new CircuitBreaker(5, 60000, Error, 'exchange_api');
registerCircuitBreaker('exchange_api', exchangeBreaker);

async function fetchExchangeData(
  exchange: string,
): Promise<Record<string, any>> {
  logger.info(`Fetching data from ${exchange}`);

  // Simulate API call
  if (Math.random() < 0.2) {
    // 20% chance of failure
    throw new DataSourceError(
      `Exchange ${exchange} API error`,
      exchange,
      ErrorCode.DATA_SOURCE_UNAVAILABLE,
      true,
    );
  }

  return { exchange, price: 100.0 };
}

async function fetchExchangeWithBreaker(
  exchange: string,
): Promise<Record<string, any>> {
  return exchangeBreaker.call(() => fetchExchangeData(exchange));
}

// Example 3: Using circuit breaker decorator
class DataRetriever {
  private breaker: CircuitBreaker;

  constructor() {
    this.breaker = new CircuitBreaker(3, 30000, Error, 'data_service');
    registerCircuitBreaker('data_service', this.breaker);
  }

  async fetchWithRetry(source: string): Promise<Record<string, any>> {
    return retry(() => this.breaker.call(() => this.fetchData(source)), {
      maxAttempts: 3,
      baseDelay: 500,
    });
  }

  private async fetchData(source: string): Promise<Record<string, any>> {
    logger.info(`Fetching from ${source}`);

    // Simulate API call
    if (Math.random() < 0.25) {
      // 25% chance of failure
      throw new DataSourceError(
        `Failed to fetch from ${source}`,
        source,
        ErrorCode.DATA_SOURCE_UNAVAILABLE,
        true,
      );
    }

    return { source, data: 'example' };
  }
}

// Example 4: Graceful degradation with fallback
async function fetchPrimaryData(): Promise<Record<string, any>> {
  logger.info('Fetching from primary source');

  // Simulate failure
  if (Math.random() < 0.5) {
    throw new DataSourceError(
      'Primary source unavailable',
      'primary',
      ErrorCode.DATA_SOURCE_UNAVAILABLE,
      true,
    );
  }

  return { source: 'primary', data: [1, 2, 3] };
}

async function fetchFallbackData(): Promise<Record<string, any>> {
  logger.info('Using fallback source');
  return { source: 'fallback', data: [4, 5, 6] };
}

async function fetchWithFallbackExample(): Promise<Record<string, any>> {
  return withFallback(
    fetchPrimaryData,
    fetchFallbackData,
    (error) => error instanceof DataSourceError,
  );
}

// Example 5: Using timeout wrapper
async function slowOperation(): Promise<string> {
  logger.info('Starting slow operation');
  await new Promise((resolve) => setTimeout(resolve, 5000));
  return 'completed';
}

async function fetchWithTimeout(): Promise<string> {
  return withTimeout(slowOperation, 2000); // 2 second timeout
}

// Example 6: Combining multiple patterns
async function robustFetch(url: string): Promise<Record<string, any>> {
  // Combine retry, circuit breaker, timeout, and fallback
  const breaker = new CircuitBreaker(3, 30000, Error, 'robust_fetch');

  const primaryFetch = async () => {
    return retry(
      () =>
        breaker.call(() =>
          withTimeout(
            () => fetchDataFromAPI(url),
            5000, // 5 second timeout
          ),
        ),
      {
        maxAttempts: 3,
        strategy: RetryStrategy.EXPONENTIAL,
        baseDelay: 1000,
      },
    );
  };

  const fallbackFetch = async () => {
    logger.warn('Using cached data as fallback');
    return { source: 'cache', data: [] };
  };

  return withFallback(primaryFetch, fallbackFetch);
}

// Main function to run examples
async function main() {
  console.log('=== Resilience Patterns Examples ===\n');

  // Example 1: Retry
  console.log('1. Retry with exponential backoff:');
  try {
    const result = await fetchWithRetry('https://api.example.com/data');
    console.log('   Success:', result, '\n');
  } catch (error) {
    console.log('   Failed:', error, '\n');
  }

  // Example 2: Circuit breaker
  console.log('2. Circuit breaker:');
  try {
    const result = await fetchExchangeWithBreaker('binance');
    console.log('   Success:', result, '\n');
  } catch (error) {
    console.log('   Failed:', error, '\n');
  }

  // Example 3: Combined retry and circuit breaker
  console.log('3. Combined retry and circuit breaker:');
  const retriever = new DataRetriever();
  try {
    const result = await retriever.fetchWithRetry('external_api');
    console.log('   Success:', result, '\n');
  } catch (error) {
    console.log('   Failed:', error, '\n');
  }

  // Example 4: Graceful degradation
  console.log('4. Graceful degradation with fallback:');
  try {
    const result = await fetchWithFallbackExample();
    console.log('   Success:', result, '\n');
  } catch (error) {
    console.log('   Failed:', error, '\n');
  }

  // Example 5: Timeout
  console.log('5. Timeout wrapper:');
  try {
    const result = await fetchWithTimeout();
    console.log('   Success:', result, '\n');
  } catch (error) {
    console.log('   Failed: Operation timed out\n');
  }

  // Example 6: Robust fetch with all patterns
  console.log('6. Robust fetch (all patterns combined):');
  try {
    const result = await robustFetch('https://api.example.com/robust');
    console.log('   Success:', result, '\n');
  } catch (error) {
    console.log('   Failed:', error, '\n');
  }
}

// Run examples if this file is executed directly
if (require.main === module) {
  main().catch(console.error);
}

export {
  fetchWithRetry,
  fetchExchangeWithBreaker,
  DataRetriever,
  fetchWithFallbackExample,
  fetchWithTimeout,
  robustFetch,
};

File: packages/agents/src/examples/resilience_example.py
"""
Example usage of resilience patterns (retry and circuit breaker).

This module demonstrates how to use retry logic and circuit breakers
to handle failures in external service calls.
"""
import asyncio
import random
from typing import Dict, Any

from ..resilience import (
    retry,
    RetryStrategy,
    CircuitBreaker,
    circuit_breaker,
    withFallback,
    register_circuit_breaker,
)
from ..errors import DataSourceError, ErrorCode
from ..logging import get_logger

logger = get_logger(__name__)


# Example 1: Using retry decorator with exponential backoff
@retry(max_attempts=3, strategy=RetryStrategy.EXPONENTIAL, base_delay=1.0)
def fetch_data_from_api(url: str) -> Dict[str, Any]:
    """
    Fetch data from an external API with automatic retry.
    
    This function will automatically retry up to 3 times with exponential backoff
    if the API call fails.
    """
    logger.info(f"Fetching data from {url}")
    
    # Simulate API call that might fail
    if random.random() < 0.3:  # 30% chance of failure
        raise DataSourceError(
            "API temporarily unavailable",
            source=url,
            code=ErrorCode.DATA_SOURCE_UNAVAILABLE,
            retryable=True,
        )
    
    return {"status": "success", "data": [1, 2, 3]}


# Example 2: Using circuit breaker decorator
@circuit_breaker(failure_threshold=5, recovery_timeout=60.0, name="exchange_api")
def fetch_exchange_data(exchange: str) -> Dict[str, Any]:
    """
    Fetch data from exchange API with circuit breaker protection.
    
    If the exchange API fails 5 times, the circuit breaker will open
    and reject requests for 60 seconds to allow the service to recover.
    """
    logger.info(f"Fetching data from {exchange}")
    
    # Simulate API call
    if random.random() < 0.2:  # 20% chance of failure
        raise DataSourceError(
            f"Exchange {exchange} API error",
            source=exchange,
            code=ErrorCode.DATA_SOURCE_UNAVAILABLE,
            retryable=True,
        )
    
    return {"exchange": exchange, "price": 100.0}


# Example 3: Combining retry and circuit breaker
class DataRetriever:
    """Example class using both retry and circuit breaker patterns"""
    
    def __init__(self):
        # Create circuit breaker for external service
        self.breaker = CircuitBreaker(
            failure_threshold=3,
            recovery_timeout=30.0,
            name="data_service",
        )
        register_circuit_breaker("data_service", self.breaker)
    
    @retry(max_attempts=3, base_delay=0.5)
    def fetch_with_retry(self, source: str) -> Dict[str, Any]:
        """Fetch data with retry logic"""
        return self.breaker.call(self._fetch_data, source)
    
    def _fetch_data(self, source: str) -> Dict[str, Any]:
        """Internal method to fetch data"""
        logger.info(f"Fetching from {source}")
        
        # Simulate API call
        if random.random() < 0.25:  # 25% chance of failure
            raise DataSourceError(
                f"Failed to fetch from {source}",
                source=source,
                code=ErrorCode.DATA_SOURCE_UNAVAILABLE,
                retryable=True,
            )
        
        return {"source": source, "data": "example"}


# Example 4: Graceful degradation with fallback
def fetch_primary_data() -> Dict[str, Any]:
    """Fetch data from primary source"""
    logger.info("Fetching from primary source")
    
    # Simulate failure
    if random.random() < 0.5:
        raise DataSourceError(
            "Primary source unavailable",
            source="primary",
            code=ErrorCode.DATA_SOURCE_UNAVAILABLE,
            retryable=True,
        )
    
    return {"source": "primary", "data": [1, 2, 3]}


def fetch_fallback_data() -> Dict[str, Any]:
    """Fetch data from fallback source"""
    logger.info("Using fallback source")
    return {"source": "fallback", "data": [4, 5, 6]}


def fetch_with_fallback() -> Dict[str, Any]:
    """Fetch data with automatic fallback"""
    return withFallback(
        fetch_primary_data,
        fetch_fallback_data,
        fallback_condition=lambda e: isinstance(e, DataSourceError),
    )


# Example 5: Async retry
@retry(max_attempts=3, base_delay=0.5)
async def fetch_async_data(url: str) -> Dict[str, Any]:
    """Async function with retry logic"""
    logger.info(f"Async fetching from {url}")
    
    # Simulate async API call
    await asyncio.sleep(0.1)
    
    if random.random() < 0.3:
        raise DataSourceError(
            "Async API error",
            source=url,
            code=ErrorCode.DATA_SOURCE_UNAVAILABLE,
            retryable=True,
        )
    
    return {"url": url, "data": "async result"}


def main():
    """Run examples"""
    print("=== Resilience Patterns Examples ===\n")
    
    # Example 1: Retry
    print("1. Retry with exponential backoff:")
    try:
        result = fetch_data_from_api("https://api.example.com/data")
        print(f"   Success: {result}\n")
    except Exception as e:
        print(f"   Failed: {e}\n")
    
    # Example 2: Circuit breaker
    print("2. Circuit breaker:")
    try:
        result = fetch_exchange_data("binance")
        print(f"   Success: {result}\n")
    except Exception as e:
        print(f"   Failed: {e}\n")
    
    # Example 3: Combined retry and circuit breaker
    print("3. Combined retry and circuit breaker:")
    retriever = DataRetriever()
    try:
        result = retriever.fetch_with_retry("external_api")
        print(f"   Success: {result}\n")
    except Exception as e:
        print(f"   Failed: {e}\n")
    
    # Example 4: Graceful degradation
    print("4. Graceful degradation with fallback:")
    try:
        result = fetch_with_fallback()
        print(f"   Success: {result}\n")
    except Exception as e:
        print(f"   Failed: {e}\n")
    
    # Example 5: Async retry
    print("5. Async retry:")
    try:
        result = asyncio.run(fetch_async_data("https://api.example.com/async"))
        print(f"   Success: {result}\n")
    except Exception as e:
        print(f"   Failed: {e}\n")


if __name__ == "__main__":
    main()
